packages/ckeditor5-engine/src/conversion/downcasthelpers.js

```diff
@@ -1670,6 +1896,247 @@ function prepareDescriptor( highlightDescriptor, data, conversionApi ) {
+// @param {String} model.name The name of element.
+// @param {Array.<String>} model.attributes The list of attribute names that should trigger reconversion.
+// @param {Boolean} [model.children] Whether the child list change should trigger reconversion.
+// @param {Function} The callback for checking a single diff item whether it should trigger reconversion.
+// @returns {Function}
+function createChangeReducer( model, callback = createChangeReducerCallback( model ) ) {
+	return ( evt, data ) => {
+		const reducedChanges = [];
+
+		if ( !data.reconvertedElements ) {
+			data.reconvertedElements = new Set();
+		}
+
+		for ( const change of data.changes ) {
+			// For attribute use node affected by the change.
+			// For insert or remove use parent element because we need to check if it's added/removed child.
+			const node = change.position ? change.position.parent : change.range.start.nodeAfter;
+
+			const elements = callback( node, change );
+
+			if ( !elements ) {
+				reducedChanges.push( change );
+
+				continue;
+			}
+
+			for ( const element of elements ) {
+				// If it's already marked for reconversion, so skip this change.
+				if ( data.reconvertedElements.has( element ) ) {
+					continue;
+				}
+
+				data.reconvertedElements.add( element );
+
+				const position = ModelPosition._createBefore( element );
+
+				reducedChanges.push( {
+					type: 'remove',
+					name: element.name,
+					position,
+					length: 1
+				}, {
+					type: 'reinsert',
+					name: element.name,
+					position,
+					length: 1
+				} );
+			}
+		}
+
+		data.changes = reducedChanges;
+	};
+}
+
+// Creates a function that checks if an element and it's watched attributes can be consumed and consumes them.
+//
+// @param {Object} model A normalized `config.model` converter configuration.
+// @param {String} model.name The name of element.
+// @param {Array.<String>} model.attributes The list of attribute names that should trigger reconversion.
+// @param {Boolean} [model.children] Whether the child list change should trigger reconversion.
+// @returns {module:engine/conversion/downcasthelpers~ConsumerFunction}
+function createConsumer( model ) {
+	return ( node, consumable ) => {
+		const events = [ 'insert' ];
+
+		// Collect all set attributes that are triggering conversion.
+		for ( const attributeName of model.attributes ) {
+			if ( node.hasAttribute( attributeName ) ) {
+				events.push( `attribute:${ attributeName }` );
+			}
+		}
+
+		if ( !events.every( event => consumable.test( node, event ) ) ) {
+			return false;
+		}
+
+		return events.every( event => consumable.consume( node, event ) );
+	};
+}
+
+// Creates a function that create view slots.
+//
+// @param {module:engine/model/element~Element} element
+// @param {Map.<module:engine/view/element~Element,Array.<module:engine/model/node~Node>>} slotsMap
+// @param {module:engine/conversion/downcastdispatcher~DowncastConversionApi} conversionApi
+// @returns {Function}
```

Comment:
I am missing here at least a sentence about the parameter which the function can take.